### PR TITLE
Add SampleNode stats and cost rule

### PR DESCRIPTION
**Pull request by Lewuathe**
The cost of `SampleNode` can be calculated by multiplying sampling ratio.

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/cost/CostCalculatorUsingExchanges.java [hunk 1]
```diff
@@ -302,6 +303,13 @@ public PlanCostEstimate visitUnion(UnionNode node, Void context)
             return costForStreaming(node, LocalCostEstimate.zero());
         }
 
+        @Override
+        public PlanCostEstimate visitSample(SampleNode node, Void context)
+        {
+            LocalCostEstimate localCost = LocalCostEstimate.ofCpu(getStats(node).getOutputRowCount());
+            return costForSource(node, localCost);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/cost/CostCalculatorUsingExchanges.java]: `costForSource` could perhaps be named "cost for leaf". It assumes (and, I think, enforces) that node has no sources, which is not the case for SampleNode

Comment (Lewuathe) [reply presto-main/src/main/java/io/prestosql/cost/CostCalculatorUsingExchanges.java]: Thanks. But I'm still not sure which to use `costForStreaming` and `costForAccumulation`.  I guess `costForAccumulation` is an estimation for aggregation type operator so we can use `costForStreamin`.

**Comment by findepi** [comment]
`SampleNode` has two types: SYSTEM, BERNOULLI.

It's hard to say what's the (execution) cost of BERNOULLI sampling, because it's never executed. Instead, it's replaced with a Filter (`ImplementBernoulliSampleAsFilter`).
Since the Filter is based on `rand()`, current code (`FilterStatsCalculator`) won't be able to estimate it properly, but maybe we need some special path just for this case.
Question is whether this replacement happens before or after cost-based decisions. If it happens after cost-based decisions, we may want to change that.

SYSTEM sampling is tricky, because its cumulative cost can be lower than the cumulative cost of its source.
https://github.com/prestosql/presto/blob/aba0c3ff34017b04949acdd0281af8da6c821f9b/presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java#L1188
When `SampleNode`'s source is `TableScan` you should be able to scale down source cost by sample's ratio multiplier -- only some splits will be processed, not all.
Not sure what to do when source is something else. For example, with aggregation (with no grouping keys) -- all the work needs to be done (sampling won't change anything here), so sampling cumulative cost will be (roughly?) the same as source's cumulative cost.

**Comment by tooptoop4** [comment]
is this related to https://github.com/prestodb/presto/pull/13452 ?

**Comment by raunaqmorarka** [comment]
Already implemented by https://github.com/trinodb/trino/pull/8094

**Pull request closed without merging**